service/geopmdpy_test/TestRuntime.py

```diff
@@ -0,0 +1,237 @@
+#!/usr/bin/env python3
+#
+#  Copyright (c) 2015 - 2021, Intel Corporation
+#
+#  Redistribution and use in source and binary forms, with or without
+#  modification, are permitted provided that the following conditions
+#  are met:
+#
+#      * Redistributions of source code must retain the above copyright
+#        notice, this list of conditions and the following disclaimer.
+#
+#      * Redistributions in binary form must reproduce the above copyright
+#        notice, this list of conditions and the following disclaimer in
+#        the documentation and/or other materials provided with the
+#        distribution.
+#
+#      * Neither the name of Intel Corporation nor the names of its
+#        contributors may be used to endorse or promote products derived
+#        from this software without specific prior written permission.
+#
+#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+#  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+#  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+#  A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+#  OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+#  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+#  LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+#  DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+#  THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+#  (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY LOG OF THE USE
+#  OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+#
+
+import unittest
+import itertools
+import math
+from unittest import mock
+from time import time
+
+# Patch dlopen to allow the tests to run when there is no build
+with mock.patch('cffi.FFI.dlopen', return_value=mock.MagicMock()):
+    from geopmdpy.runtime import TimedLoop
+    from geopmdpy.runtime import Agent
+    from geopmdpy.runtime import Controller
+
+class PassthroughAgent(Agent):
+    def __init__(self, signals, controls, period):
+        self._signals = signals
+        self._controls = controls
+        self._period = period
+        self._adjust_value = 7
+        self._report_data = 'report data'
+
+    def get_signals(self):
+        return self._signals
+
+    def get_controls(self):
+        return self._controls
+
+    def get_period(self):
+        return self._period
+
+    def update(self, signals):
+        return [self._adjust_value] * len(signals)
+
+    def get_report(self):
+        return self._report_data
+
+class TestRuntime(unittest.TestCase):
```

Comment:
This works, but I'd rather have the test classes map to the classes they are testing.  It means we have to have multiple test files for runtime.py, TestTimedLoop.py and TestController.py, and the tests for TimedLoop and Controller would be independent.